### PR TITLE
104046 patch punch - vol1 - Input validation

### PR DESCRIPTION
**Pull request by janid1967**
[AB#104046](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/104046) 

The PR has many file changes due to `<Nullable>enable</Nullable>` in test project for WebApi
Added several ToDo's marked 104046 ... more PR's to come to solve patching of punch

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.WebApi/Controllers/PatchOperationValidator.cs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+                }
+            }
+        }
+
+        return illegalOperations;
+    }
+
+    private static bool CanAssignValueToProperty(object? value, Type propType)
+    {
+        if (value == null)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.WebApi/Controllers/PatchOperationValidator.cs:128]: value is null

File: src/Equinor.ProCoSys.Completion.WebApi/Controllers/PatchOperationValidator.cs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+            return IsNullableType(propType);
+        }
+
+        var valueType = value.GetType();
+        if (propType.IsAssignableFrom(valueType))
+        {
+            return true;
+        }
+
+        if ((propType == typeof(double) || propType == typeof(double?)) && value is int)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.WebApi/Controllers/PatchOperationValidator.cs:140]: So, we're not supporting double, only int?

File: src/Equinor.ProCoSys.Completion.WebApi/Controllers/PatchOperationValidator.cs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        }
+
+        var replaceStringOperations = operations
+            .Where(op => op.OperationType == OperationType.Replace && op.value is string);
+        foreach (var operation in replaceStringOperations)
+        {
+            var propName = operation.path.TrimStart('/');
+            var propertyWithLengthLimiting =
+                propertiesWithLengthLimiting.SingleOrDefault(p => p.Name == propName);
+            if (propertyWithLengthLimiting == null)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.WebApi/Controllers/PatchOperationValidator.cs:243]: is null

File: src/Equinor.ProCoSys.Completion.WebApi/Controllers/PatchOperationValidator.cs (hunk 2 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.WebApi/Controllers/PatchOperationValidator.cs:140]: Line 140 is special case of assigning an int to a double-property. (I'll add comment)
Assign int to int-property is handled in line 135: propType.IsAssignableFrom(valueType)

(line 135 is false for assigning an int to a double-property)

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Controllers/PatchOperationValidatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+        // Act
+        var result1 = _dut.AllRequiredFieldsHaveValue(_patchDocument.Operations);
+        var result2 = _dut.GetMessageForRequiredFields(_patchDocument.Operations);
+
+        // Assert
+        Assert.IsFalse(result1);
+        Assert.IsNotNull(result2);
+        Assert.IsTrue(result2.EndsWith("MyDateTime"));
+    }
+
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Controllers/PatchOperationValidatorTests.cs:601]: Missing AllRequiredFieldsHaveValue_ShouldReturnFalse_WhenTryingToSetNullInRequiredDouble ?

**Review by jsoi-eq — APPROVED** [review]

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Controllers/PatchOperationValidatorTests.cs (hunk 4 above)
Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Controllers/PatchOperationValidatorTests.cs:623]: added

**Pull request merged**